Adding support for Private DNS zones Public Preview

muwaqar opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

This PR is loosely based on https://github.com/Azure/azure-powershell-pr/pull/294 which was submitted when the Private DNS zones feature was launched in a private preview. The PR keeps most of the changes.

It is built on top of 2018-03-01-preview Swagger version (https://github.com/Azure/azure-rest-api-specs/pull/2587) and version 2.2.0-preview of the .NET SDK (https://github.com/Azure/azure-sdk-for-net/pull/4109). These changes are not merged/live yet, so this PR is just to initiate the design review and get an approval in principle. Once the pre-requisite changes have been merged, only then we plan to complete this PR.

The relevant cmdlet design review for this PR can be found at https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/22

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#making-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

maddieclayton commented:
@muwaqar Do you have cmdlet review associated with this PR?  If not, please fill out a review here: https://github.com/Azure/azure-powershell-cmdlet-review-pr.  Once this is approved, we will proceed with the PR review.

maddieclayton commented:
@muwaqar To be able to properly build without a published nuget package, you will need to put your unpublished nuget package here: https://github.com/Azure/azure-powershell/tree/preview/tools/LocalFeed

muwaqar commented:
@maddieclayton : We are waiting for the Swagger and .NET SDK PRs to complete (mentioned in the description). Once they get merged, I will push another commit here with the final .NET SDK version.

cormacpayne commented:
@maddieclayton what's the status of this PR?

maddieclayton commented:
@cormacpayne Still waiting for SDK to be merged and published.  @muwaqar please assign this to me once the SDK is published and the PR is updated.

muwaqar commented:
@maddieclayton : I generated the MSI (https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/posh-pack/52/). Now, because of `PSVirtualNetwork` in New/Set-AzureRmDnsZone cmdlets, we have a dependency on AzureRm.Network module. Because of which, I have to do `Import-Module AzureRm.Network` before `Import-Module AzureRm.Dns`, otherwise using New/Set-AzureRmDnsZone cmdlet fails saying that the Network DLL was not found. Is there a way to automate this so that importing Dns module always imports Network module? Or is it only because the MSI was delay signed (instead of strong signed).

cormacpayne commented:
@muwaqar we don't want to have cross-module dependencies -- in this case, we have a `Commands.Common.Network` project that contains generated code from the Network SDK that you should be using instead of the types exposed from the Network module. I think we need to look at creating a common `PSVirtualNetwork` type that is stored in `Commands.Common.Network` and is used by both your module and the Network module, so users don't have to load the `AzureRM.Network` module anytime they want to use your commands.

maddieclayton commented on src/ResourceManager/Dns/Commands.Dns/Commands.Dns.csproj:
```diff
@@ -115,6 +121,10 @@
       <Project>{3819d8a7-c62c-4c47-8ddd-0332d9ce1252}</Project>
       <Name>Commands.ResourceManager.Common</Name>
     </ProjectReference>
+    <ProjectReference Include="..\..\Network\Commands.Network\Commands.Network.csproj">
```
Use this project instead: https://github.com/Azure/azure-powershell/blob/2001d5053dfda2cbc3d9a04ba826b042311d6964/src/ResourceManager/Compute/Commands.Compute/Commands.Compute.csproj#L343

maddieclayton commented on src/ResourceManager/Dns/Commands.Dns/SetEnv.ps1:
```diff
@@ -0,0 +1,21 @@
+﻿Param(
```
What is this and where is it used?

muwaqar replied:
It's a helper file that we use to record the tests. Without it, it's very cumbersome to set it up.

maddieclayton replied:
This setup method already exists and is described here: https://github.com/Azure/azure-powershell/blob/preview/documentation/testing-docs/using-azure-test-framework.md.  Please remove this ps1 file.

maddieclayton commented on src/ResourceManager/Dns/Commands.Dns/Zones/NewAzureDnsZone.cs:
```diff
@@ -12,20 +12,28 @@
 
 namespace Microsoft.Azure.Commands.Dns
 {
+    using System.Linq;
```
Move these above

maddieclayton commented on src/ResourceManager/Dns/Commands.Dns/Zones/NewAzureDnsZone.cs:
```diff
@@ -35,10 +43,30 @@ public class NewAzureDnsZone : DnsBaseCmdlet
         public string ResourceGroupName { get; set; }
 
+        [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The type of the zone, Public or Private. This property cannot be changed for a zone.")]
+        [ValidateNotNullOrEmpty]
```
Please add a validateset here if the only optional allow are public and private.

maddieclayton commented on src/ResourceManager/Dns/Commands.Dns/Zones/SetAzureDnsZone.cs:
```diff
@@ -53,17 +80,46 @@ public override void ExecuteCmdlet()
-                    Etag = "*",
-                    Tags = this.Tag,
-                };
+                    this.Name = this.Name.TrimEnd('.');
```
Would this have previously caused the cmdlet to fail?

muwaqar replied:
This check was already present in the file, on line 76.

maddieclayton commented on tools/StaticAnalysis/Exceptions/SignatureIssues.csv:
```diff
@@ -1087,3 +1087,11 @@
 "Microsoft.Azure.Commands.AnalysisServices.dll","Microsoft.Azure.Commands.AnalysisServices.NewAzureRmAnalysisServicesFirewallRule","New-AzureRmAnalysisServicesFirewallRule","1","8100","New-AzureRmAnalysisServicesFirewallRule Does not support ShouldProcess but the cmdlet verb New indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it should imp ...
 "Microsoft.Azure.Commands.AnalysisServices.dll","Microsoft.Azure.Commands.AnalysisServices.NewAzureRmAnalysisServicesFirewallConfig","New-AzureRmAnalysisServicesFirewallConfig","1","8100","New-AzureRmAnalysisServicesFirewallConfig Does not support ShouldProcess but the cmdlet verb New indicates that it should.","Determine if the cmdlet should implement ShouldProcess and if so determine if it shou ...
 "Microsoft.WindowsAzure.Commands.Storage.dll","Microsoft.WindowsAzure.Commands.Storage.Common.Cmdlet.EnableAzureStorageServiceDeleteRetentionPolicyCommand","Enable-AzureStorageDeleteRetentionPolicy","1","8410","Parameter RetentionDays of cmdlet Enable-AzureStorageDeleteRetentionPolicy does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a  ...
+"Microsoft.Azure.Commands.Dns.dll","Microsoft.Azure.Commands.Dns.SetAzureDnsZone","Set-AzureRmDnsZone","1","8410","Parameter RegistrationVirtualNetworkIds of cmdlet Set-AzureRmDnsZone does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
```
Please change all of the "VirtualNetworkIds" to "VirtualNetworkId" and all of the "VirtualNetworks" to "VirtualNetwork".  This is a powershell standard.  Once you do that you can remove these lines from SignatureIssues.

muwaqar commented on src/ResourceManager/Dns/Commands.Dns.Test/Commands.Dns.Test.csproj:
```diff
@@ -63,7 +63,12 @@
+    </Reference>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=17.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
+      <SpecificVersion>False</SpecificVersion>
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.17.0.0-preview\lib\net452\Microsoft.Azure.Management.Network.dll</HintPath>
```
I don't get this. Can you clarify?

maddieclayton commented:
@muwaqar Besides the Network issue, this PR looks good to me.

cormacpayne commented on src/ResourceManager/Dns/AzureRM.Dns.psd1:
```diff
@@ -54,7 +54,7 @@ CLRVersion = '4.0'
 
 # Assemblies that must be loaded prior to importing this module
-RequiredAssemblies = '.\Microsoft.Azure.Management.Dns.dll'
+RequiredAssemblies = '.\Microsoft.Azure.Management.Dns.dll', '.\Microsoft.Azure.Commands.Network.dll'
```
@muwaqar this reference to `Microsoft.Azure.Commands.Network.dll` can be removed

cormacpayne commented on src/ResourceManager/Dns/Commands.Dns.Test/Commands.Dns.Test.csproj:
```diff
@@ -63,7 +63,12 @@
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Dns.2.2.0-preview\lib\net452\Microsoft.Azure.Management.Dns.dll</HintPath>
+      <Private>True</Private>
+    </Reference>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=17.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
@muwaqar if your test project is going to depend on the Network SDK, then you should update the Network entry in the [`TestMappings.json`](https://github.com/Azure/azure-powershell/blob/preview/TestMappings.json#L111-L117) file with a reference to the Dns test project so these tests get run whenever a change in Network is made

cormacpayne commented on src/ResourceManager/Dns/Commands.Dns/Commands.Dns.csproj:
```diff
@@ -89,13 +90,17 @@
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Dns.2.0.0-preview\lib\net452\Microsoft.Azure.Management.Dns.dll</HintPath>
+      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Dns.2.2.0-preview\lib\net452\Microsoft.Azure.Management.Dns.dll</HintPath>
+    </Reference>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=17.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
@muwaqar this reference shouldn't be needed anymore, right?

cormacpayne commented on src/ResourceManager/Network/Commands.Network/Models/PSVirtualNetwork.cs:
```diff
@@ -14,11 +14,12 @@
+    using Microsoft.Azure.Management.Internal.Network.Common;
+    using Newtonsoft.Json;
+    
+    public class PSVirtualNetwork : PSTopLevelResource, IResourceReference
```
@muwaqar since we're making changes in the Network module, we will need to release a new version of their module along with Dns so users are able to access this updated type. Our mechanism for determining if a module has been updated or not is an entry in the change log, so you will need to update the Network change log with a short snippet about updating model types for compatibility with Dns cmdlets.

muwaqar commented:
@cormacpayne: Fixed the issues for which you posted comments.
Regarding the .NET SDK, our PR (https://github.com/Azure/azure-sdk-for-python/pull/2132) has been completed and just awaiting nuget generation. I will update this PR as soon as we have a nuget and remove the one I added to the LocalFeed for now. I understand this PR cannot be completed before then, but we want to reach a state where we have a sign-off on all other aspects of the PR besides the nuget change.

cormacpayne commented on src/ResourceManager/Dns/Dns.sln:
```diff
@@ -36,6 +36,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Common.Network", "
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Common.Storage", "..\..\Common\Commands.Common.Storage\Commands.Common.Storage.csproj", "{65C3A86A-716D-4E7D-AB67-1DB00B3BF72D}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Commands.Network", "..\Network\Commands.Network\Commands.Network.csproj", "{98CFD96B-A6BC-4F15-AE2C-603FC2B58981}"
```
@muwaqar you should include the `Commands.Common.Network` project in this solution as well so it gets built with your solution. Also, do you need to build `Commands.Network` with this solution, or can we remove it?

muwaqar replied:
`Commands.Common.Network` is already part of this solution.

muwaqar commented:
@maddieclayton @cormacpayne : I have updated PR with published .NET SDK nuget. The build seems to be failing due to some unrelated issue. Please re-queue the build, and if you don't have any further comments, please merge it at your earliest convenience.

cormacpayne commented:
On-demand: https://azuresdkci.westus2.cloudapp.azure.com/job/powershell-demand/606/
